### PR TITLE
Misc TypeScript fixes

### PR DESCRIPTION
**Pull request by tikotzky**
This PR introduces a few updates to the TypeScript setup and typings.

Reviewing type changes can sometimes be confusing for me, so in attempt to make it easier to review I made each change in its own commit. If you have questions or would like any changes let me know.

I'll explain the changes in each commit in a more detail  below.
_**Note:** none of the changes should break any existing types, all the changes should be making existing types more correct and better defined._

#### move config to a `tsconfig.json` file and enable strict mode
The switch to a `tsconfig.json` was just made to get matching config settings when using an IDE.
Enabling strict mode is to fix and prevent problems when using this fastify from a within a project with strict mode enabled.

#### add types for `cors` library
Once strict mode is enabled it wants types for all imports, so we add the types for `cors` as a dev dependency

#### add overload to allow `.use` to accept express middleware
The current types only allow passing fastify middleware which was giving a type error when trying to use `cors` (since `cors` is a type of `express.RequestHandler`). This commit adds a overload to allow passing either fastify middleware or express middleware to `use`

#### always pass a callback to middleware
another strict mode fix:)
The types for the callback specified that the callback would be optionally passed to middleware.
_(types currently are `callback?: (err?: Error) => void` note the `?` after callback)_
This means that in strict mode you would never be able to call `next` directly, since it would be possibly `undefined`. 
For example this was giving a types error
```ts
app.addHook('onRequest', (req, res, next) => {
	next(); // next is possibly 'undefined'.
});
```
This changes the type to specify that the callbacks will always be passed.
Note: This does not mean that all middleware must take a callback middleware.
This will still work as expected after this change
```ts
app.addHook('onRequest', (req, res) => {
	res.end('Hello World');
});
```

#### default and constrain generic types
When falling back to the generic instance of `fastify()` if you did not specify generic params then it would end up being a generic of `<{}, {}, {}>`. This means that the below case give type errors.
```ts
// current definition
declare function fastify<HttpServer, HttpRequest, HttpResponse>...

const app = fastify({ logger: true });
// app is now a FastifyInstance<{}, {}, {}>
app.addHook('onRequest', (req, res) => {
	res.end('Hello World'); // Property 'end' does not exist on type '{}'.
});

// defaulting generics
declare function fastify<
  HttpServer = http.Server, 
  HttpRequest = http.IncomingMessage, 
  HttpResponse = http.ServerResponse
>...
const app = fastify({ logger: true });
// app is now a FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse>
```

as for constraining generics, it requires that the generic being passed in extends from the base type whic we expect...
```ts
// types were allowing this...
const app = fastify<number, string, boolean>({ logger: true });

//constrain generics
declare function fastify<
  HttpServer extends http.Server
  HttpRequest extends http.IncomingMessage
  HttpResponse extends http.ServerResponse
>
// the is no longer valid
const app = fastify<number, string, boolean>({ logger: true });
```

Putting together defaults and constraints looks like 
```ts
declare function fastify<
  HttpServer extends http.Server = http.Server, 
  HttpRequest extends http.IncomingMessage = http.IncomingMessage, 
  HttpResponse extends http.ServerResponse = http.ServerResponse
>...
```

_**fin**_ 
Sorry that this got a bit wordy, I figured its better to explain than to just open a PR with a bunch of  changes to the types 😄

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -97,6 +98,7 @@
     "x-xss-protection": "^1.0.0"
   },
   "dependencies": {
+    "@types/express": "^4.11.0",
```
Comment (mcollina) [inline package.json]: This is not needed. We do not depend on Express.

File: fastify.d.ts [hunk 2]
```diff
@@ -6,18 +6,23 @@ import * as http from 'http';
...
 import * as https from 'https';
 import * as pino from 'pino';
+import * as express from 'express';
 
-declare function fastify<HttpServer, HttpRequest, HttpResponse>(opts?: fastify.ServerOptions): fastify.FastifyInstance<HttpServer, HttpRequest, HttpResponse>;
+declare function fastify<
+  HttpServer extends http.Server = http.Server, 
+  HttpRequest extends http.IncomingMessage = http.IncomingMessage, 
+  HttpResponse extends http.ServerResponse = http.ServerResponse
+>(opts?: fastify.ServerOptions): fastify.FastifyInstance<HttpServer, HttpRequest, HttpResponse>;
```
Comment (mcollina) [inline fastify.d.ts]: We also support Http2. Not sure how to specify this here.

File: fastify.d.ts [hunk 3]
```diff
@@ -261,11 +266,13 @@ declare namespace fastify {
      * Apply the given middleware to all incoming requests
      */
     use(middleware: Middleware<HttpServer, HttpRequest, HttpResponse>): void
+    use(middleware: express.RequestHandler): void
```
Comment (mcollina) [inline fastify.d.ts]: This is not correct :(.

Comment (delvedor) [inline fastify.d.ts]: We do not expose the enhanced version of the request/response object of express.
Is this correct?

Comment (tikotzky) [reply fastify.d.ts]: I agree about this... The `cors` library currently types itself as express middleware, so I added that really just to keep the tests passing.
I'll work on removing the dep on express...

File: package.json (hunk 1 above)
Comment (tikotzky) [reply package.json]: I'll work on removing it.

File: fastify.d.ts (hunk 3 above)
Comment (tikotzky) [reply fastify.d.ts]: gonna try to remove it :)

File: fastify.d.ts (hunk 2 above)
Comment (tikotzky) [reply fastify.d.ts]: This is just defaulting the case where you did not pass any generic type.
http2 support is enabled below based on wether or not the `http2` option was set to `true`

I dont think it should affect http2 let me know if you have a test case which its breaking and I'll work on fixing it,

Comment (mcollina) [reply fastify.d.ts]: I'm grand with how it is. I do not know how to fix it - I wrote my first piece of TS yesterday and it was __very hard__.

Comment (tikotzky) [reply fastify.d.ts]: You were having trouble with TS and fastify? Or in general?

Comment (mcollina) [reply fastify.d.ts]: In general, I even fixed a bug :D. https://github.com/fastify/fastify/pull/736

Comment (tikotzky) [reply fastify.d.ts]: cool 👍 

I've been using TS for a little while, but am now trying to see what it would take for use to use fastify for a new project.

Thats what driving me to fix up the types 😄

**Comment by tikotzky** [comment]
So I dug into this a bit more, currently the `cors` module is typing itself as an `express.RequestHandler` (even though they really work with more than express)

As of now the options as I see it regarding this PR
1. leave the PR as is, adding "compatibility" with express middleware in as advertised [here](https://www.fastify.io/docs/latest/Middlewares/)
1. remove express and work on fixing the test for the `cors` middleware.

If we choose option 2 there is two paths we can take.
1. fix up the `cors` module before merging this PR (this would make me sad since it means you would not be able to use this lib in TS strict mode)
1. manually specify the type of `cors` module to `any` in the test, which doesn't test anything. And that brings me to the third choice of
1. remove the test for `cors` since without `cors` fixing their types this test is not testing anything.

Please let me know how'd you like me to proceed...

**Comment by tikotzky** [comment]
I pushed up a change removing the dep on express and changing the `import` or `cors` in the test to use `require`.
Switching to require causes `cors` to be an `any` which means that its really not testing anything.

Just wanted to push something for now. Let me know if you want any more changes...

File: test/types/index.ts [hunk 4]
```diff
@@ -2,12 +2,13 @@
 // This file will be passed to the TypeScript CLI to verify our typings compile
 
 import * as fastify from '../../fastify'
-import * as cors from 'cors'
 import * as http from 'http';
 import * as http2 from 'http2';
 import { readFileSync } from 'fs'
 import { createReadStream, readFile } from 'fs'
 
+const cors = require('cors');
```
Comment (mcollina) [inline test/types/index.ts]: Can you add a comment on why this is done in this way?

File: fastify.d.ts [hunk 5]
```diff
@@ -7,17 +7,21 @@ import * as http2 from 'http2';
...
+>(opts?: fastify.ServerOptions): fastify.FastifyInstance<HttpServer, HttpRequest, HttpResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsHttp): fastify.FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsSecureHttp): fastify.FastifyInstance<https.Server, http.IncomingMessage, http.ServerResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsHttp2): fastify.FastifyInstance<http2.Http2Server, http2.Http2ServerRequest, http2.Http2ServerResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsSecureHttp2): fastify.FastifyInstance<http2.Http2SecureServer, http2.Http2ServerRequest, http2.Http2ServerResponse>;
 
 declare namespace fastify {
 
-  type Plugin<HttpServer, HttpRequest, HttpResponse, T> = (instance: FastifyInstance<HttpServer, HttpRequest, HttpResponse>, opts: T, callback?: (err?: Error) => void) => void
+  type Plugin<HttpServer, HttpRequest, HttpResponse, T> = (instance: FastifyInstance<HttpServer, HttpRequest, HttpResponse>, opts: T, callback: (err?: Error) => void) => void
```
Comment (mcollina) [inline fastify.d.ts]: the callback is optional. If you are not using it, you are supposed to return a promise (or declare it as an async function).

File: test/types/index.ts (hunk 4 above)
Comment (tikotzky) [reply test/types/index.ts]: I'll add a comment but this test is not really doing anything at this point since `cors` is typed as an `any`...

File: fastify.d.ts (hunk 5 above)
Comment (tikotzky) [reply fastify.d.ts]: my understanding is that using the callback is optional but the callback itself is always passed to the plugin/middleware.

In any case I'll try add an overload omitting the callback if you return a promise.

Comment (tikotzky) [reply fastify.d.ts]: Can both middleware and plugins return promises?
I can't seem to find docs regarding returning promises.

Comment (mcollina) [reply fastify.d.ts]: plugins and hooks and handlers can return a promise or support async/await

File: fastify.d.ts (hunk 2 above)
Comment (evanshortiss) [reply fastify.d.ts]: @tikotzky these will break http2 type support: `HttpServer extends http.Server`. We support passing in the generics so correct typings can be determined later in handlers etc. and this includes http2 typings. http2 doesn't use the same structure as http so we can't enforce that extension.

Comment (evanshortiss) [reply fastify.d.ts]: Perhaps a union type can solve this? For example:

```
declare function fastify<
   HttpServer extends http.Server | HttpServer extends http2.Server
   // etc
 >
```

Although it seems like http2 does not export the server class - https://github.com/nodejs/node/blob/master/lib/http2.js

Comment (tikotzky) [reply fastify.d.ts]: syntax for a union type would be
```js
declare function fastify<
  HttpServer extends (http.Server | http2.Http2Server) = http.Server, 
  HttpRequest extends (http.IncomingMessage | http2.Http2ServerRequest) = http.IncomingMessage, 
  HttpResponse extends (http.ServerResponse | http2.Http2ServerResponse) = http.ServerResponse
>...
```

and it does seem to work.
Gonna push that and a test for it too...

File: fastify.d.ts (hunk 5 above)
Comment (tikotzky) [reply fastify.d.ts]: ok, so first off there is no special support for async/await. async/await is just syntactical sugar over promises. From the perspective of the type system async function are the same as regular functions which return a promise.

As for adding the overload, I can add an overload but there doesn't seem to be any gain in doing so.

The current type looks something like
```ts
cb(req: HttpRequest, res: HttpResponse, callback: (err?: Error) => void) => void
```
Typescript does not enforce that a function use all params that its passed.
In addition when the return type is set to void TS ignore the return value as well.
What this means is that all these are valid for the current definition that gets passed a callback.
```ts
server.use(() => {})
server.use((req) => {})
server.use((req, res) => {})
server.use((req, res, next) => {})
server.use(() => {return anything})
server.use((req) => {return anything})
server.use((req, res) => {return anything})
server.use((req, res, next) => {return anything})
server.use(async () => {})
server.use(async (req) => {})
server.use(async (req, res) => {})
server.use(async (req, res, next) => {})
server.use(async () => {return anything})
server.use(async (req) => {return anything})
server.use(async (req, res) => {return anything})
server.use(async (req, res, next) => {return anything})
```

If we were to add an additional overload that looked like
```ts
cb(req: HttpRequest, res: HttpResponse) => Promise<void>
```
It would not protect us from returning a promise and taking in next.
Reason being that since the type signature of the callback variant allows for functions which return promise (because of the `void` return type), as soon as we accept a 3rd param to our function it will just switch to using the callback signature and will not complain about using next.

Comment (mcollina) [reply fastify.d.ts]: use does not support promises. register does.

Comment (mcollina) [reply fastify.d.ts]: what would be good to support is that, if you are not receiving next, you have to return a promise.

Comment (mcollina) [reply fastify.d.ts]: I would prefer to leave that as optional.

File: test/types/index.ts [hunk 6]
```diff
@@ -2,12 +2,15 @@
...
 
 import * as fastify from '../../fastify'
-import * as cors from 'cors'
 import * as http from 'http';
 import * as http2 from 'http2';
 import { readFileSync } from 'fs'
 import { createReadStream, readFile } from 'fs'
 
+// were importing cors using require, which causes it to be an `any`. This is done because `cors` exports
+// itself as an express.RequestHandler which is not compatible with the fastify TypeScript types
```
Comment (mcollina) [inline test/types/index.ts]: Could we cast it to a compatible middleware interface that we define in fastify?

File: test/types/index.ts [hunk 7]
```diff
@@ -30,12 +33,28 @@ import { createReadStream, readFile } from 'fs'
...
+  })
+
+  interface CustomHttp2IncomingMessage extends http2.Http2ServerRequest {
+    getDeviceType: () => string;
+  }
+
+  const customHttp2Server: fastify.FastifyInstance<http2.Http2Server, CustomHttp2IncomingMessage, http2.Http2ServerResponse> = fastify();
+  customHttp2Server.use((req, res, next) => {
+    console.log('can access props from CustomIncomingMessage', req.getDeviceType());
+  });
```
Comment (mcollina) [inline test/types/index.ts]: why are you adding this test?

File: test/types/index.ts (hunk 6 above)
Comment (tikotzky) [reply test/types/index.ts]: we can cast to whatever we want, but it does not prove anything.
If we cast, all we prove is that the cast is a correct cast.
But does not prove that the lib works correctly with middleware.

If you wanted to cast you can test by casting any function, there would be no benefit to importing and casting `cors`

File: test/types/index.ts (hunk 7 above)
Comment (tikotzky) [reply test/types/index.ts]: This is testing that you can create set the generic types to be an `http2` server.
The `use` and `log` are ensuring that the generic request type is passed along when executing `.use((req, res, next) => {})`

Comment (mcollina) [reply test/types/index.ts]: Thanks!

File: test/types/index.ts (hunk 6 above)
Comment (mcollina) [reply test/types/index.ts]: @evanshortiss what do you think?

**Comment by mcollina** [comment]
@evanshortiss can this be landed?

**Comment by corbinu** [comment]
Anything I can do to help with this?

File: fastify.d.ts (hunk 5 above)
Comment (nwoltman) [reply fastify.d.ts]: I think all that's needed is to add `| Promise` to the very end here.

**Comment by mcollina** [comment]
@corbinu do you think this is ready to land?

**Comment by corbinu** [comment]
@mcollina I cloned @tikotzky 's version, re-based on master and the tests were failing due to implicit type any in a few places in the test script. 

I also pulled it into my own Fastify project and that compiled fine 😄 

However the failures can be fixed (I can't seem to comment directly as the lines aren't currently changed in the PR though if I am just missing something please let me know)

in `test/types/index.ts`
Line 116: 
```
const opts = {
```

should be 

```
const opts: fastify.RouteShorthandOptions<http2.Http2Server, http2.Http2ServerRequest, http2.Http2ServerResponse> = {
```

at the bottom 
```
server.ready(function (err, done) {
  done(err)
})

server.ready(function (err, context, done) {
  server.log.debug(context)
  done(err)
})

server.ready()
  .then((context) => {
    server.log.debug(context)
  })
  .catch((err) => {
    server.log.error(err)
  })

server.after(function (err) {
  if (err) throw err
})

server.after(function (err, done) {
  done(err)
})

server.after(function (err, context, done) {
  server.log.debug(context)
  done(err)
})
```
should be
```
server.ready(function (err: Error, done: Function): void {
  done(err)
})

server.ready(function (err: Error, context: fastify.FastifyInstance<http2.Http2Server, http2.Http2ServerRequest, http2.Http2ServerResponse>, done: Function) {
  server.log.debug(context)
  done(err)
})

server.ready()
  .then((context) => {
    server.log.debug(context)
  })
  .catch((err) => {
    server.log.error(err)
  })

server.after(function (err) {
  if (err) throw err
})

server.after(function (err: Error, done: Function) {
  done(err)
})

server.after(function (err: Error, context: fastify.FastifyInstance<http2.Http2SecureServer, http2.Http2ServerRequest, http2.Http2ServerResponse>, done: Function) {
  server.log.debug(context)
  done(err)
})
```

Once it is in I can do a follow on PR for some improvements I would like to see.

**Comment by mcollina** [comment]
I do not understand we need to have:

```
FastifyInstance<http2.Http2SecureServer, http2.Http2ServerRequest, http2.Http2ServerResponse>
```

Can't we just have `FastifyInstance` everywhere (as it is now)?

**Comment by corbinu** [comment]
FastifyInstance takes generics so they need to be specified to determine if it is http or http2. However I agree it is verbose I plan to add some short cuts in my own PR. That’s simply what is required to make the tests pass on this PR

**Comment by tikotzky** [comment]
I just rebased and pushed up the changes needed to get this PR in a passing state...

**Comment by corbinu** [comment]
@tikotzky Great!!!! tests pass great for me now. Thank you so much for all this work!

File: test/types/index.ts (hunk 6 above)
Comment (evanshortiss) [reply test/types/index.ts]: @tikotzky @mcollina I don't think there's much benefit to casting either. Could there be some benefit to changing the type signature of `fastify.use` to accept `express.RequestHandler` instances?

File: test/types/index.ts [hunk 8]
```diff
@@ -256,11 +275,11 @@ server.ready(function (err) {
   if (err) throw err
 })
 
-server.ready(function (err, done) {
+server.ready(function (err: Error, done: Function) {
   done(err)
 })
 
-server.ready(function (err, context, done) {
+server.ready(function (err: Error, context: fastify.FastifyInstance<http2.Http2SecureServer, http2.Http2ServerRequest, http2.Http2ServerResponse>, done: Function) {
```
Comment (evanshortiss) [inline test/types/index.ts]: Is it necessary to specify that `done` is a function? I assume the types are added here to demonstrate passing a specific context i.e HTTP2?

Comment (corbinu) [reply test/types/index.ts]: The types are being added so that tests pass in strict mode. So it can either be type `any` or `Function`

File: test/types/index.ts (hunk 6 above)
Comment (mcollina) [reply test/types/index.ts]: @evanshortiss they are not the same. Some middlewares are compatible, but not all. Specifically, the types of `cors` are wrong as that should be a standard http response from core, not an Express one.

Comment (evanshortiss) [reply test/types/index.ts]: Aah, you're correct. Maybe something to revisit in a separate issue/pr?
